aptos-move/framework/aptos-framework/sources/stake.move

```diff
@@ -632,22 +654,33 @@ module aptos_framework::stake {
             error::invalid_argument(EALREADY_ACTIVE_VALIDATOR),
         );
 
-        // Requires the validator to have a lockup more than the minimum.
         let config = staking_config::get();
         let (minimum_stake, maximum_stake) = staking_config::get_required_stake(&config);
-
-        // The validator is not yet activated so all added stake should be in active.
-        let voting_power = coin::value<AptosCoin>(&stake_pool.active);
+        let voting_power = get_next_epoch_voting_power(stake_pool);
         assert!(voting_power >= minimum_stake, error::invalid_argument(ESTAKE_TOO_LOW));
         assert!(voting_power <= maximum_stake, error::invalid_argument(ESTAKE_TOO_HIGH));
 
+        // Track and validate voting power increase.
+        // Except for if the validator set's total voting power is 0, which is the case for genesis.
+        let validator_set = borrow_global_mut<ValidatorSet>(@aptos_framework);
+        let total_voting_power = validator_set.total_voting_power;
+        if (total_voting_power > 0) {
+            validator_set.total_joining_power = validator_set.total_joining_power + (voting_power as u128);
+            validate_voting_power_increase(
+                validator_set.total_joining_power,
+                total_voting_power,
+                staking_config::get_voting_power_increase_limit(&config),
+            );
```

Comment:
this code is consistently the same, I would change the function to just take in the increase amount, get the mut resource, update it, evaluate it, etc...

Reply:
I was trying to optimize gas here by not calling borrow_global_mut from multiple times. But I guess this is something we can fix on the gas schedule side - borrow_global/borrow_global_mut of the same resource should not charge a lot multiple times as the resource is already cached after the first call.